src/app/MessageDef.h

```diff
@@ -0,0 +1,894 @@
+     */
+    CHIP_ERROR GetReaderOnTag(const uint64_t aTagToFind, chip::TLV::TLVReader * const apReader) const;
+
+protected:
+    chip::TLV::TLVReader mReader;
+
+    ParserBase();
+
+    template <typename T>
+    CHIP_ERROR GetUnsignedInteger(const uint8_t aContextTag, T * const apLValue) const;
+
+    template <typename T>
+    CHIP_ERROR GetSimpleValue(const uint8_t aContextTag, const chip::TLV::TLVType aTLVType, T * const apLValue) const;
+};
+
+class ListParserBase : public ParserBase
+{
+protected:
+    ListParserBase();
+
+public:
+    // aReader has to be on the element of the array element
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
+
+    // aReader has to be at the beginning of some container
+    CHIP_ERROR InitIfPresent(const chip::TLV::TLVReader & aReader, const uint8_t aContextTagToFind);
+
+    CHIP_ERROR Next();
+    void GetReader(chip::TLV::TLVReader * const apReader);
+};
+
+class BuilderBase
+{
+public:
+    void ResetError();
+    void ResetError(CHIP_ERROR aErr);
+    CHIP_ERROR GetError() const { return mError; };
+    chip::TLV::TLVWriter * GetWriter() { return mpWriter; };
+
+protected:
+    CHIP_ERROR mError;
+    chip::TLV::TLVWriter * mpWriter;
+    chip::TLV::TLVType mOuterContainerType;
+
+    BuilderBase();
+    void EndOfContainer();
+
+    CHIP_ERROR InitAnonymousStructure(chip::TLV::TLVWriter * const apWriter);
+};
+
+class ListBuilderBase : public BuilderBase
+{
+protected:
+    ListBuilderBase();
+
+public:
+    /**
+     * Init the TLV array container with an anonymous tag.
+     * Required to implement arrays of arrays, and to test ListBuilderBase.
+     * There is no chip interaction model message that has an array as the outermost container.
+     *
+     * @param[in]   apWriter    Pointer to the TLVWriter that is encoding the message.
+     * @param[in]   aContextTagToUse    A contextTag to use.
+     *
+     * @return                  CHIP_ERROR codes returned by Chip::TLV objects.
+     */
+
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter, const uint8_t aContextTagToUse);
+    /**
+     * Init the TLV array container with an anonymous tag.
+     * Required to implement arrays of arrays, and to test ListBuilderBase.
+     * There is no chip interaction model message that has an array as the outermost container.
+     *
+     * @param[in]   apWriter    Pointer to the TLVWriter that is encoding the message.
+     *
+     * @return                  CHIP_ERROR codes returned by Chip::TLV objects.
+     */
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter);
+};
+
+namespace AttributePath {
+enum
+{
+    kCsTag_NodeId              = 0,
+    kCsTag_EndpointId          = 1,
+    kCsTag_NamespacedClusterId = 2,
+    kCsTag_FieldId             = 3,
+    kCsTag_ListIndex           = 4,
+};
+
+class Parser : public ParserBase
+{
+public:
+    // aReader has to be on the element of AttributePath container
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
```

Comment:
is there any reason that `Init()` isn't on the base class?

Reply:
AttributePath, EventPath, CommandPath, they all inherit from BuildBase, in init, they start the container with  kTLVType_Path.  StartContainer(aTag, chip::TLV::kTLVType_Path, mOuterContainerType)

class ListBuilderBase(inherited by AttributePathList..) also inherits from BuilderBase, but it start the container with kTLVType_Array 
mpWriter->StartContainer(chip::TLV::AnonymousTag, chip::TLV::kTLVType_Array, mOuterContainerType);

therefore, I am not putting the init in BuildBase, maybe we can create another ArrayBuildBase to have this common array init?

Reply:
I see we've addressed this comment for BuilderBase, why not for ParserBase?

Reply:
Done, thanks